### DISCUSSION TITLE
Printing inner and outer walls at the same speed by default

### DISCUSSION THREAD

**Discussion by jellespijker** [Ultimaker/Cura#8974] [Ideas]
> _Note_:  This discussion was opened on behave of @jmding8  whom posted some interesting findings in the FB group: Cura Ultimaker 3D Slicer User Group. See original post [here](https://www.facebook.com/groups/2207556406016768/permalink/3162556950516704/) 

# Original question

I'm having trouble with Cura and seams. The picture shows the same model sliced with Cura 4.8 on the left, and PrusaSlicer on the right. Both were sliced using the 0.20mm speed profile for the Mk3s, and printed on a stock Kingroon KP3s (direct drive model). The model is a simple, 20mm x 20mm cylinder, with 0% infill and 2 perimeters. Note, I'm using the Mk3s profile that you have to download from Prusa's website and import manually to Cura. For both slicers, I set seam position to random to emphasize the problem. As you can see from the photo (taken with aggressive lighting to emphasize the issue), the Cura print shows prominent overextrusion for ~10mm after the seam (the nozzle moves from left to right) resulting in an exaggerated layer-line artifact. While the PrusaSlicer version also has slight extrusion inconsistencies around the seam, it is significantly less prominent and the zit itself is much smaller.

![131905590_4118842443437_6300688347463384184_o](https://user-images.githubusercontent.com/8535734/102689804-4b42c680-4201-11eb-98b9-0e0ae4f001b6.jpg)

I'm more trying to understand why this is happening than looking for a solution. While I could minimize the issue by reducing forcing a retraction to occur when moving from inner to outer walls, and then adding a negative retraction extra prime amount, I'm really just looking to understand the cause of this effect.

I've tried turning on retract after layer change, turning on optimize wall printing order (which makes the outer perimeter seam line up with the inner perimeter seam so there's basically no travel required when transitioning from inner to outer wall), and reducing jerk and acceleration, none of this helped. I looked into the gcode as well, thinking that maybe Cura was overextruding at the start of the line or that PrusaSlicer was underextruding at the start, but the extrusion distance per XY distance was basically constant for both. I even printed the same files on my Ender3, which has an Hermera, which is about as responsive an extrusion setup as possible, but it shows the same issue, so it doesnt seem like an extrusion hardware problem. I'm out of ideas at this point. Anyone know what's going on?

# Findings

After a TON of investigation (20 test prints last night alone...) I FINALLY figured out what the cause of these lines is! You can see the original Cura print on the left, the fixed Cura print in the middle, and the PrusaSlicer print on the right. TLDR: the fix is to just print inner and outer walls at the same speed, instead of printing inner walls faster (as many profiles, and indeed the Cura defaults, often do).

If you want to understand why, including enough background to know why the fix I just mentioned above WONT always work, feel free to read on. Please forgive me for being long-winded (and a bit patronizing even?), I wanted to make this explanation thorough for all audiences (and I'm a bad writer).

The faster you extrude plastic, the more pressure you need in the hotend. Note that when I say "faster," I don't mean your nozzle speed (typically something like 50mm/s), I mean how much volume of plastic your extruder/hotend is flowing out the nozzle. For example, printing 50mm/s, with a 0.4mm line width, and a 0.2mm layer height results in 4mm^3/s of plastic flow (50mm/s * 0.4mm * 0.2mm). That is also the same volumetric flow rate as printing 100mm/s, with a 0.4mm line width, and a 0.1mm layer height (100mm/s * 0.4mm * 0.1mm). This should make some intuitive sense, if you want your toothpaste to come out of the tube faster, you have to squeeze harder.

But let's say you need to suddenly and dramatically reduce your volumetric flow rate. For example, let's say you're printing a 0.2mm layer, with a 0.4mm line width, at 100mm/s (8mm^3/s volumetric flow), and want to suddenly drop to 50mm/s without changing layer height or line width (4mm^3/s volumetric flow). That sudden halving of volumetric flow would mean you need to just as suddenly cut the pressure in your hotend by about half as well. It's not easy to do that, because all the extra pressure it took to print at a high flow rate of 8mm^3/s takes time to dissipate and causes a little bit of oozing out of the nozzle as it gradually goes away.

So hw does that apply here? Well, we tend to print the outer walls more slowly than the rest of the print in order to reduce ghosting and ringing artifacts. So if a printer goes from printing infill, or even an inner wall, at a relatively high flow rate, then suddenly goes to print an outer wall at a much slower flow rate, it needs to (almost instantaneously) drop the pressure in the nozzle which it simply cannot do. All that extra pressure results in extra material being extruded that the firmware and gcode are not accounting for. That extra material shows up as a bulging line immediately after the seam.
Why does this not happen with PrusaSlicer? It's because Cura is pretty unique among slicers in that it allows the user to set a different print speeds for the outer wall vs the inner walls. In fact the default 0.2mm "draft" profile seems to have 60mm/s inner wall speed and 30mm/s outer wall speed. This forces the printer to go from printing the inner walls at a high flow rate, to printing the outer wall at half that flow rate, with basically no time to allow the pressure in the nozzle to dissipate first. That extra pressure results in the bulging line. PrusaSlicer meanwhile doesn't have this feature, so inner and outer walls print at the same speed, and so the volumetric flow and nozzle pressure are both nice and constant and smooth when moving from the inner walls to the outer wall.

Technically actually, you can produce the same problem with PrusaSlicer, you just have to try kind of hard. You have to set the infill to print faster than the walls, set the outer wall to print before the inner walls, AND print a small object. This would force the printer to rapidly transition from high flow infill extrusion to low flow outer wall extrusion, resulting in extra pressure in the hotend. And if your object is small as well, the travel move from infill to wall would be short, and so that extra pressure wouldn't have time to dissipate and would cause a bulging line. This doesn't happen in practice very often though, because the way PrusaSlicer is set up makes it unlikely that you'd end up doing this accidentally.

So what can we, as Cura users, do about it? The most practical solution is to just print inner and outer walls at the same speed. Simple. There's a bunch of other stuff that can help too, like turning on linear advance (maybe... Marlin's implementation might not be aggressive enough for this situation, more testing needed), significantly reducing jerk and acceleration, forcing a travel between inner and outer walls to give time for the extra pressure and material to ooze out as stringing, force a retraction between inner and outer walls and use a negative retraction extra prime distance, etc etc. But most of these other solutions have major drawbacks and are only useful situationally.

![131954238_4119497739819_6412628428742280932_o (1)](https://user-images.githubusercontent.com/8535734/102689818-6e6d7600-4201-11eb-8a1b-f92b7fd83b1d.jpg)

**Comment by jellespijker** [comment]
@jmding8 could you explain why you think that increasing the speed of the molten polymer in the liquefier and die, result in an increase of pressure?
Because in my experience it's the other way around. Are you familiar with Bernoulli's principle ( I certainly hope so seeing that you're a mechanical engineer 😉) ( https://en.m.wikipedia.org/wiki/Bernoulli%27s_principle ) this principle basically says that the sum of the kinetic, static and potential energy at point N along the fluid lines is constant. If we assume for arguments sake that a molten polymer behaves as Newtonian fluid (e.q. water), which is actually not the case (since it behaves as a Bingham plastic or thixotropic fluid) and we also assume that it is incomprehensible we can eliminate the specific density from both the kinetic and potential. This principle simply states that if you increase on energy term you have to decrease an other. Meaning that an increase in speed will result in a pressor drop or vice versa.
Just think about it. Energy can't be destroyed only converted. So in order to have the sum of energy's the same, increasing the speed v in the equation <img src="https://render.githubusercontent.com/render/math?math=\frac{v^2}{2}%2Bgz%2B\frac{p}{\rho}=const.">. Means that you have to decrease the pressure p.
Or do you have other effects in mind why that might be the case?

**Reply by jellespijker** [reply]
Since you indicated that your prime goal is to understand why this is happening I thought this might be an interesting read for you. Last year I made a simulation model for what is actually happening in the liquefier and die. They're two interlinked models. One is a flow model as seen on the left, while the right one is a thermodynamic model.

![liquefier-model-v1](https://user-images.githubusercontent.com/8535734/102690043-0881ee00-4203-11eb-8aae-3a8a9c6c9ec5.jpg)

#  Material Feed
The system is fed a certain volume-metric flow of solid filament. Which is done by rotating the gripper wheels of the feeder at a certain speed. This rotational speed, coupled with the cross section of the filament and the density of the material at atmospheric pressure and ambient temperature, obtained from a 2-domain p-v-T relation material property, translates to a mass flow of solid filament being fed into the liquefier.

![CodeCogsEqn (3)](https://user-images.githubusercontent.com/8535734/102690125-80501880-4203-11eb-8496-b49ed58fd5b2.gif)

![CodeCogsEqn (2)](https://user-images.githubusercontent.com/8535734/102690134-91992500-4203-11eb-9a49-ba29230793a2.gif)

![CodeCogsEqn (5)](https://user-images.githubusercontent.com/8535734/102690140-9fe74100-4203-11eb-91cc-fc4792393c3b.gif)

![CodeCogsEqn (4)](https://user-images.githubusercontent.com/8535734/102690148-b1304d80-4203-11eb-965b-f6267ce59555.gif)

# Heat flow
Each time step the temperature of the system is modeled and fed into the melt and flow processes. The material properties rely heavily on the temperature of the material and these determine the behaviour of the extrusion and melting process.

The assumption is made that the liquefier is going to be a radial systems, with a known ambient temperature.

This thermal simulation model is modeled as a graph which is a mathematical structure used to model pairwise relations between objects. A graph in this context is made up of vertices, nodes, or points which are connected by edges, arcs, or lines. 

All transference of energy flows through edges. These could therefor be seen as resistor which is a passive two-terminal electrical component that implements electrical resistance as a circuit elements. Where a node represent a point, such as a compartment or, transition between layers. This approach is an analogy to an electrical network.

For each node in a graph the summation of in flowing heat, is equal to the outgoing heat plus the internal energy change. 

![CodeCogsEqn (7)](https://user-images.githubusercontent.com/8535734/102690162-d45afd00-4203-11eb-9a4e-133b14c04bc0.gif)

While an edge represent an resistance to the transference of heat from node to node. Since heat flow is thermal potential difference divide by thermal resistance, it can be expressed as follows:

![CodeCogsEqn (8)](https://user-images.githubusercontent.com/8535734/102690185-ff455100-4203-11eb-985c-3fdf5f0266d6.gif)

from this the resistance can  be expressed as:

![CodeCogsEqn (9)](https://user-images.githubusercontent.com/8535734/102690208-21d76a00-4204-11eb-94ea-db42791d43fd.gif)

The resistance from heat transfer to due to conduction (for instance in a wall) can be expressed in a radial system as:

![CodeCogsEqn (10)](https://user-images.githubusercontent.com/8535734/102690218-3156b300-4204-11eb-87f0-ad19b0aaf7a5.gif)

here r_0 is the outside radius and r_i he inner radius, k is the thermal conductivity property of the material and L the length of the annulus.

Convection boundary condition of the chamber in which the liquefier resides can be described with:

![CodeCogsEqn (11)](https://user-images.githubusercontent.com/8535734/102690234-45021980-4204-11eb-8b21-90a5f34bb5e9.gif)

Here h is the convection heat-transfer coefficient.
> Note: The heat transfer resistance due to convection in a pseudoplastic moving in a laminar flow regime is still to be determined, but the current assumptions is that this will be the same as the normal convection resistance.

Whereas the heat transfer resistance due to radiation can be expressed as:

![CodeCogsEqn (12)](https://user-images.githubusercontent.com/8535734/102690260-695df600-4204-11eb-8c1e-d18c34a9d81c.gif)

Here sigma_c is the Boltzmann constant, epsilon 1 and 2 are the material emissivity properties and F_12 is the view factor between the two surfaces.

# Phase change
This solid filament undergoes a phase change due to the heat flowing into the system. Which goes from the heaters, through the wall and then either through air (assumed to be none moving) or the minniscus (with a dynamic height) into the solid filament. Where it will heat up the solid filament, until it undergoes a phase change. In which the temperature of the filament no longer rises until it is completely melted. Afterwards it will continue to heat up until it has reached the desired extrusion temperature. From where it will be treated as flow domain problem.

> Note: Due to the complexities and uncertainties which arise at this melting process it will be modeled as a simple time delay, between feed rate of the solid filament and flow rate of the liquefied filament. The exact model of this delay has yet to be determined. But it will most likely be build on the basic principal of: 
![CodeCogsEqn](https://user-images.githubusercontent.com/8535734/102690279-8bf00f00-4204-11eb-956f-5bcf10b81c1c.gif)

# Liquid flow
The actual extrusion process is modeled as a parallel pipe system, consisting of a main flow pipe, just after the melt process, branching of in to the minniscus and the extrusion path, containing a reducer and a die. The melt of the filament behaves like a pseudoplastic and can therefor be treated as a power law fluid. The following calculations are commonly used for the design of plastic extrusion books.
The relation between flow rate and pressure drop can be expressed with the equation below. Here V_dot is the previously obtained volume-metric flow rate, G is a die constant Delta p is the pressure difference, K is a factor of proportionality and n is the power law exponent.

![CodeCogsEqn (13)](https://user-images.githubusercontent.com/8535734/102690293-afb35500-4204-11eb-9afd-2e45392242c2.gif)

K and n can be determined with the Herschel-Bulkey model for non-Newtonian fluids:

![CodeCogsEqn (14)](https://user-images.githubusercontent.com/8535734/102690297-bfcb3480-4204-11eb-9ca0-8eb080b67bc7.gif)

which can be rewritten in:

![CodeCogsEqn (15)](https://user-images.githubusercontent.com/8535734/102690312-cfe31400-4204-11eb-8e25-17a2793559fb.gif)

From rheological measurements the K and n can be obtained using a simple regression line.  Since the apparent dynamic viscosity is defined as:

![CodeCogsEqn (16)](https://user-images.githubusercontent.com/8535734/102690322-dffaf380-4204-11eb-9db5-6d66f9477384.gif)

The apparent dynamic viscosity can be calculated with:

![CodeCogsEqn (17)](https://user-images.githubusercontent.com/8535734/102690330-f86b0e00-4204-11eb-8587-1188d1b3dfa3.gif)

The shear rate for a circle is given as:

![CodeCogsEqn (18)](https://user-images.githubusercontent.com/8535734/102690341-06b92a00-4205-11eb-96f2-aa037203a98d.gif)

and the shear rate for an annulus (minniscus):

> Note: This doesn't take into account the moving wall of the filament

The die constant or G for a circular shape is:

![CodeCogsEqn (20)](https://user-images.githubusercontent.com/8535734/102690370-3ec06d00-4205-11eb-80af-9fe957ce6da1.gif)

while the die constant G for an annulus is :

![CodeCogsEqn (21)](https://user-images.githubusercontent.com/8535734/102690380-4bdd5c00-4205-11eb-876e-71876a04d5b3.gif)

Rewrite the first equation into:

![CodeCogsEqn (22)](https://user-images.githubusercontent.com/8535734/102690390-5bf53b80-4205-11eb-8b98-a0ca65ae5c7b.gif)

gives the pressure drop over that segment.

# Used sources
* N. S. Rao, Design Formulas for Plastics Engineering. Munich?; New York?: New York: Hanser Gardner Publications, 1991.
* N. S. Rao, Basic Polymer Engineering Data. Cincinnati, Ohio, USA: Hanser, 2017.
* S. Kainth, Die Design for Extrusion of Pipes and Tubes: A Practical Guide. Munich?: Hanser Publications?; Cincinnati: Hanser, 2017.
* C. Hopmann and W. Michaeli, Extrusion Dies for Plastics and Rubber: Design and Engineering Computations. Carl Hanser Verlag GmbH & Company KG, 2017. 
* M. Neut, “FDM process brief (draft V1).” .
* R. W. Johnson, Handbook of Fluid Dynamics. CRC Press, 2016.
* J. H. Lienhard, A Heat Transfer Textbook. Courier Corporation, 2011.
* G. A. SCHRAMM, A Practical Approach to Rheology and Rheometry. Haake, 1994.
* J. P. Holman, Heat Transfer (Si Units). McGraw-Hill, 2008

**Reply by jmding8** [reply]
This is amazing! I read through it, and understood only a fraction, but will continue to review it.

For the purposes of the original discussion however, I think the very last equation is all we need:

![image](https://user-images.githubusercontent.com/44815547/102696753-6e5f7d00-41e5-11eb-8db5-a1b97d4c3b03.png)

If V* volumetric flow increases, delta_p increases, since K, G and n are constants.

We know that the pressure of the fluid at the nozzle is equal to ambient air pressure (an uncontained liquid cannot be pressurized), so if delta_p increases, then pressure inside the meltzone must have increased as well.

**Reply by jellespijker** [reply]
@jmding8 Unfortunatly K isn't a constant since K is depended on the the rheological properties of the polymer. In particular the shear rate which is depended on the velocity and the viscosity of the fluid. Where the viscosity of the fluid is heavenly influenced by temperature.
![CodeCogsEqn (9)](https://user-images.githubusercontent.com/8535734/102751441-0dac7d80-4368-11eb-848a-5319c06d2242.gif)
The same goes for n.
Although G might be considered constant in most parts of the model. We still have to account for an adjustment of G at the height of the miniscus. Where the inner wall of the annulus is not a static but in all actuality a downwards moving rod (filament being fed into the system).

**Reply by THeijmans** [reply]
Polymer physics are difficult, Polymer melts are very non-newtonian and quite compressible... What you are seeing here is the compressibility and system response of everything between the feeder wheels and the nozzle die. That includes deformation of the filament, melt level in the nozzle, compression of the melt and deformation fo the extruder itsself.
But we see similar effects in our prints. It's a lot more visible on a direct drive, as it has a flow response of ~0.5-1.5 seconds. That means that it will reach the requested flow ~20mm after the z-seam when the print speed is 20mm/s on the outer wall. At the z-seam, it's still at the previous flow setpoint. 

The z-seam itsself has a negative effect on it as it stops the system (depending on the jerk/acceleration parameters). 

Features like linear advance can remove this effect. But it's also better to reduce the difference between your flow (/print speed) setpoints.

**Reply by jellespijker** [reply]
@THeijmans Have you guys ever investigated how the relationship between speed of the inner and outer wall correlate to surface blobs?

**Reply by jmding8** [reply]
> It's a lot more visible on a direct drive

Thanks for mentioning this, I happened to run an experiment that is consistent with your conclusion here. Here's the exact same gcode, sliced in Cura printed on two different machines. The blue print was on an Ender3 with an Hermera extruder, where the distance from the extruder gear to the heat break is roughly 21mm, and with linear advance turned on at 0.05mm of filament compression per 1mm/s extrusion speed. The gray was printed on a Kingroon KP3s with the stock, direct drive extruder which places a Mk8 extruder directly above a hot-end that is similar to an e3d v6. The distance from extruder gear to heat break is roughly 50mm, and linear advance is disabled.

![image](https://user-images.githubusercontent.com/44815547/102925780-e11c6280-4448-11eb-83d1-20f536901314.png)

The slicing profile has outer walls at 30mm/s, inner walls at 60mm/s, 800mm/s^2 inner and outer wall acceleration, 8mm/s inner and outer wall jerk, and optimize wall printing order enabled, which minimizes the travel move from inner to outer wall and stops the printer from crossing through the cylinder to start the outer wall extrusion.

While both printers are "direct drive," the Hermera has a much shorter pressurized filament path and a faster response time. For the Hermera print, the additional compressed material leaves a very prominent, almost spherical zit at the start of the outer wall, but the excess material dissipates very quickly and by about 1/4 of the way across the frontal face of the object, the extrusion seems to have stabilized. For the Mk8 print, the excess material never forms quite as prominent of a blob, but the excess extrusion doesnt fully dissipate even after the nozzle has traversed the entire frontal face of the object. 

Interpretation is a tricky because linear advance was turned on for the Hermera printer. Still, I think it generally reinforces the observation that the effect is more prominent for direct driven printers, and also shows that linear advance is insufficient to overcome the effect of significant discontinuities in extrusion rate.


Overall, though I have yet to spend the time to verify conclusively within your theoretical model of the extrusion system that higher print speeds require higher pressures, my practical testing has left me quite convinced that there is certainly a compression effect, and that sudden changes in extrusion rate should be avoided when dealing with the outer wall. I think the ideal solution would be to gradually decrease the inner wall flow rate to match the outer wall flow rate during the last 5mm^3 of the inner wall, a shorter term solution would be to just encourage equal inner and outer wall speeds.

**Reply by THeijmans** [reply]
> @THeijmans Have you guys ever investigated how the relationship between speed of the inner and outer wall correlate to surface blobs?

Yes we have: that's the reason why we make small steps in speed between inner wall and outer wall. On a Bowden Extruder it's not as visible in a print like this (the cylinder) because the Bowden will make the flowrate for the inner wall and the outer wall roughly equal. But it can be visible on prints where the amount of material per layer changes a lot. It also creates a large dimensional accuracy problem. Having the inner wall speed double the speed of the outer walls will always lead to issues, unless there is a really good compensation feature running (either in the slicers engine or in the printer). 

Additionally, coming back to the model: because the material is non-newtonian, the viscosity goes down at higher flow rates. This breaks a lot of linearity from the simpler models... This leads to a boundary layer close to the nozzle wall where there is a high shear and very low viscosity, while the rest of the material doesn't really 'feel' the difference in flow rate.

**Comment by jellespijker** [comment]
@THeijmans & @Sarita-Ultimaker  I came across the discussion in a Facebook group and after a short chat with the author, we decided to move it here. So the material and processing team could also participate. I'm not quite sure if I agree with the conclusion Jamie Ding draws but I can't put my finger on it yet. So I was kinda hoping you guys could also take a look at it.

**Comment by Ghostkeeper** [comment]
Most of Ultimaker's print profiles have a very small speed and line width difference between the inner and outer wall for this reason. It's well understood that the heat zone acts as a sort of spring between when the feeder feeds and the nozzle reacts. This is made worse too by Ultimaker's Bowden tubes which are also like a spring which adds reaction time.

The default values for Outer Wall Speed vs. Inner Wall Speed in Custom FFF Printer are probably too far apart for most printers, yes.
There is a bit of difficulty in changing this in fdmprinter though. There are hundreds of printers and thousands of profiles that don't override these settings but are (supposedly) tuned correctly for certain speeds. If we change the default, these will be out of tune.
Basically, we could change the default in fdmprinter only if we add overrides to all profiles such that the value stays the same for the current profiles.

**Reply by jmding8** [reply]
I agree it would be difficult to change defaults and potentially "break" a large number of community profiles. In the short term, I think a reasonable solution would be to include this nugget of important, but uncommon, tribal knowledge in the tooltip next to the outer wall speed and/or inner wall speed parameter.

**Comment by BloodyRain2k** [comment]
The explaination for the pressure buildup during the inner wall and then resulting overpressure when starting the outer wall is a real enlightment to me. Even if it should've been obvious.

But given that it "is obvious" to be the reason why the speeds should match: why is there no setting for "Inner Wall Wipe Distance"?
There is "Outer Wall Wipe Distance", but that doesn't help getting rid of overpressure when transitioning from inner to outer wall.

Couldn't such a setting be used to allow using different speeds with less overpressure problems by allowing the overpressure to release in the wipe before starting the outer wall?

**Reply by jmding8** [reply]
I would argue that the better solution might be to characterize the extruder so that the slicer or firmware understands the volumetric material compression required to achieve a certain extrusion rate (basically the linear advance value), have the user state a maximum acceptable line-width error, and back out the maximum allowable extruder acceleration rate from there.

Also, if we do the inner-to-outer-wall transition "the smart way" (where the end of the inner wall is right next to the start of the outer wall, and not half way across the model), this is technically a period of acceleration. The nozzle decelerates to 0 mm/s in the tangential direction, and re-accelerates to printing speed. It also accelerates in the orthogonal direction, and then decellerates back to 0mm/s in the orgthogonal direction as it continues along the wall. But in the real world, what we'd probably prefer to do is simply neglect these accelerations, hold a constant speed in the tangential direction, and apply a velocity impulse in the orthogonal direction. This makes teh extrusion velocity as constant as possible. There ought to be special path planning logic for this special case.

**Reply by Ghostkeeper** [reply]
The Outer Wall Wipe Distance is not meant to relieve overpressure in the nozzle. After all, it's wiping on top of an outer wall. If it would be relieving pressure, that material would be deposited on top of the outer wall, resulting in a blip at the seam.

Coasting is designed to do that though. Coasting also applies to inner walls.

**Reply by Khronos666** [reply]
[Ghostkeeper](https://github.com/Ghostkeeper) The Outer Wall Wipe Distance this setting does not work in the new version of CUDA 5.5.0 if the thickness of the model is less than 3 lines... so is coasting...
even with the 5.0.0 version, I pointed to this...
